### ISSUE TITLE
Consider reordering the topics in the essentials class

### ISSUE THREAD

**Issue by sextonkennedy** [hsf-training/cpluspluscourse#482]
Hi All,
	I was a mentor for 2 sessions this past spring, once on zoom and once in person at jLab. I think you should reconsider the order of the topics for the essentials class.  I'll use the order on slide 5.  1. And 2. Are good where they are but then I would do:
6. **Useful tools** so that they can debug their exercises.
4. **Core modern C++** so that they are not tired out by the time they reach what are in my opinion some of the most important topics, const correctness, RAII, and how to **use** the STL.
3. **Object orientation** it's still great material and good to know the memory layout the compiler uses but it is lighter for the last day.

**Comment by sponce** [comment]
Usual comments on this kind of topic : 
  - very welcome if you manage but...
  - already attempted and failed. What we have is the best we've managed
  - as we say in french : "the better is the enemy of the good". Not sure the translation makes sense, but basically too much improvements usually leads to degradation rather than improvements

The main issue you will face is "dependencies" in concepts which won't be fulfilled, which will lead to "we will see that later" type of explanations, which I personally hate as a student.

But worth a try if you find the time (although there is also a long list of issue if you do have time)

**Comment by hageboeck** [comment]
It's fine for me personally. I think it's more work than you think, though, so it would have to be distributed quite well.

Just going through the first few subsections of core modern, we use (without discussing):
- member function constness
- overloading
- exceptions (which use inheritance)
- move semantics (it's part of the advanced course, so not an issue for essentials), but it's pretty OO

I think we need a little OO, but maybe it can be shortened to objects/classes, inheritance basics, constructors/destructors?
The rest of OO such as static, allocating, advanced OO, type casting, operator overloading, function objects, name lookup could be moved to a later stage.

**Comment by chavid** [comment]
Yes. Very very hard to swap parts in a such long cooked bible, without breaking many things. That's one of the reasons we tried to start a less ambitious new course from scratch, few years ago. But this died of lacking time and human resources.
